### ISSUE TITLE
SESSION-06: Inter-Row Shading Analysis with Electrical Loss Modeling

### ISSUE THREAD

**Issue by ganeshgowri-ASA** [ganeshgowri-ASA/pv-layout-designer#4]
### Objective
Calculate inter-row shading losses with bypass diode electrical modeling

### Branch
`feature/06-shading-model`

### Duration
1.5 hours

---

#### Core Functions
```python
def calculate_inter_row_shading(row_pitch: float, module_length: float, 
                                tilt_angle: float, sun_altitude: float) -> float:
    """
    Calculate shading fraction for given sun position.
    Returns 0.0 (no shade) to 1.0 (fully shaded).
    """
    
def calculate_hourly_shading(layout: Dict, date: str, lat: float, lon: float) -> List[Dict]:
    """
    Hourly shading analysis for entire day.
    Returns: [{'hour': int, 'shading_fraction': float, 'power_loss': float}]
    """
    
def calculate_electrical_loss(shading_fraction: float, 
                              bypass_diodes: int = 3) -> float:
    """
    Convert geometric shading to electrical loss using bypass diode model.
    - <5%: Linear loss
    - 33%: 1 diode bypass (~33% loss)
    - 66%: 2 diodes bypass (~66% loss)  
    - >66%: Full module loss (100%)
    """
    
def generate_shading_profile(layout: Dict, location: Dict) -> Dict:
    """
    Annual shading analysis.
    Returns: {
        'winter_solstice': {...},
        'summer_solstice': {...},
        'annual_average_loss': float
    }
    """
```

#### Algorithm Implementation
```python
def calculate_inter_row_shading(row_pitch, module_length, tilt_angle, sun_altitude):
    import numpy as np
    
    # Shadow length on ground
    shadow_length = module_length * np.sin(np.radians(tilt_angle)) / np.tan(np.radians(sun_altitude))
    
    # Module horizontal footprint
    module_footprint = module_length * np.cos(np.radians(tilt_angle))
    
    # Clear distance to next row
    clear_distance = row_pitch - module_footprint
    
    # Calculate shading
    if shadow_length > clear_distance:
        shaded_length = shadow_length - clear_distance
        shading_fraction = min(shaded_length / module_length, 1.0)
        return shading_fraction
    
    return 0.0  # No shading
```

---

### Integration with SESSION-04

```python
from src.models.solar_calculations import calculate_solar_elevation, calculate_sun_path

def calculate_hourly_shading(layout, date, lat, lon):
    sun_path = calculate_sun_path(lat, lon, date)
    
    results = []
    for hour_data in sun_path:
        if hour_data['elevation'] > 0:  # Daytime only
            shading = calculate_inter_row_shading(
                row_pitch=layout['row_pitch'],
                module_length=layout['module_length'],
                tilt_angle=layout['tilt_angle'],
                sun_altitude=hour_data['elevation']
            )
            
            electrical_loss = calculate_electrical_loss(shading)
            
            results.append({
                'hour': hour_data['hour'],
                'sun_elevation': hour_data['elevation'],
                'shading_fraction': shading,
                'electrical_loss': electrical_loss
            })
    
    return results
```

---

### Visualization in src/app.py

```python
import matplotlib.pyplot as plt
import streamlit as st

# After layout generation
if 'layout' in st.session_state:
    st.subheader("📊 Shading Analysis")
    
    # Winter solstice analysis
    date = "2024-12-21"
    shading_data = calculate_hourly_shading(
        st.session_state['layout'],
        date,
        config['latitude'],
        config['longitude']
    )
    
    # Plot
    fig, ax = plt.subplots()
    hours = [d['hour'] for d in shading_data]
    losses = [d['electrical_loss'] * 100 for d in shading_data]
    
    ax.plot(hours, losses, marker='o')
    ax.set_xlabel('Hour of Day')
    ax.set_ylabel('Power Loss (%)')
    ax.set_title('Winter Solstice Shading Losses')
    ax.grid(True)
    
    st.pyplot(fig)
    
    # Summary metrics
    col1, col2 = st.columns(2)
    col1.metric("Critical Hours Loss (9-3PM)", f"{critical_loss:.1f}%")
    col2.metric("Annual Shading Loss", f"{annual_loss:.1f}%")
```

---

## ✅ Testing Checklist

- [ ] Shadow length calculated correctly
- [ ] Shading fraction ranges 0.0-1.0
- [ ] Bypass diode model non-linear
- [ ] Winter solstice shows worst-case
- [ ] Hourly analysis returns 24 values
- [ ] Charts render without errors
- [ ] Integration with SESSION-04 works

---

## 🔄 Git Workflow

```bash
git checkout -b feature/06-shading-model

python -m py_compile src/models/shading_model.py
pytest tests/test_shading_model.py -v
streamlit run src/app.py

git add src/models/shading_model.py src/app.py tests/
git commit -m "feat(shading): inter-row shading with electrical loss modeling"
git push origin feature/06-shading-model
```

---

## 📊 Success Criteria

1. ✅ Shading calculation matches manual verification
2. ✅ Electrical losses non-linear (bypass diode effect)
3. ✅ Hourly profile shows expected pattern
4. ✅ Winter solstice worst-case validated
5. ✅ Charts display correctly in Streamlit
6. ✅ Ready for SESSION-08 visualization integration

---

## 🔗 Dependencies

- **Requires**: SESSION-04 (solar calculations), SESSION-05 (layout engine)
- **Blocks**: SESSION-08 (needs shading data for visualization)
